src/jlgen.jl

```diff
@@ -639,7 +646,13 @@ function compile_method_instance(@nospecialize(job::CompilerJob))
                 Metadata(ConstantInt(DEBUG_METADATA_VERSION()))
         end
 
-        native_code = if VERSION >= v"1.12.0-DEV.1667"
+        native_code = if VERSION >= v"1.12.0-DEV.1823"
+            # each item in the list should be a CodeInstance followed by a CodeInfo
+            # indicating something to compile
+            codeinstance = ci_cache_lookup(cache, job.source, job.world, job.world)::CodeInstance
+            codeinfo = CC.typeinf_code(interp, job.source, #=trim=# true)::CodeInfo
```

Comment:
Because we don't have easy access to the `codeinfos` array populated earlier, I'm just passing a single CodeInstance/CodeInfo pair, for the top-level call. Is that OK?

Reply:
Also, could I get the CodeInfo somewhere else, or is it fine to call `typeinf_code` again>

Reply:
Why do the worklist in ci_cache_lookup? 

I was thinking of doing it in `compile_method_instance` like https://github.com/JuliaGPU/GPUCompiler.jl/pull/582/files#diff-1bf557ada55697d453a6ccf81e9c263e46e573e324c453c4ecffb70994f36c37R689-R715

Reply:
Because it's simply the minimal change replacing our usage of `typeinf_ext_toplevel` from before. I agree it would be better to drive that differently, but that's a larger change (which I'm not sure we can do across all supported versions).

Reply:
> I'm just passing a single CodeInstance/CodeInfo pair, for the top-level call. Is that OK?

Turns out it wasn't, and we really need to pass al relevant CodeInfos to jl_emit_native. Doing so makes the test suite pass.

That said, it does require us to now always call the worklist (i.e. ci_cache_lookup, which should be renamed), even when the cache already contains a certain method instance. But I guess that should be fine, since these calls into type inference should be fast (as stuff is cached already)?